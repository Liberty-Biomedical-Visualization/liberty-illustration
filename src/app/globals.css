@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  --accent-50: 224 247 252;
  --accent-100: 178 224 240;
  --accent-200: 128 200 232;
  --accent-300: 77 176 215;
  --accent-400: 26 153 206;
  --accent-500: 3 152 206;
  --accent-600: 2 128 165;
  --accent-700: 2 109 136;
  --accent-800: 2 90 107;
  --accent-900: 2 61 77;
  --accent-950: 1 44 58;
  --background: 255 255 255;
  --foreground: 23 23 23;
}

body {
  color: var(--foreground);
  background: var(--background);
  font-family: Arial, Helvetica, sans-serif;
}

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}
