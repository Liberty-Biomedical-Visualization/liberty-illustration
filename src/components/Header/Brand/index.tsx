import Image from "next/image";

import type { ImageData } from "@/lib/content";

export default function Brand(props: BrandProps) {
  const { className, logoData } = props;
  const { description, height, src, title, width } = logoData;

  return (
    <div className={className}>
      <Image
        alt={description ?? ""}
        className="max-w-3/4 mx-auto"
        height={height}
        src={src}
        title={title}
        priority
        width={width}
      />
      <h1 className="font-light sm:text-3xl text-2xl text-center text-nowrap">
        Abigail Richbourg Liberty, MS, CMI
      </h1>
    </div>
  );
}

export interface BrandProps {
  className?: string;
  logoData: ImageData;
}
