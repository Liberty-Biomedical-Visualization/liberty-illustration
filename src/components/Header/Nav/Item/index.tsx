"use client";

import Link from "next/link";
import { usePathname } from "next/navigation";

import resolveClassNames from "@/lib/resolveClassNames";

export default function NavItem(props: Readonly<NavItemProps>) {
  const { children, href } = props;
  const pathname = usePathname();

  const color =
    pathname === href
      ? "active:text-accent-200 hover:text-accent-400 text-accent-600"
      : "active:text-accent-400 hover:text-accent-600 text-accent-800";

  const linkClassName = resolveClassNames(
    "text-lg transition uppercase",
    color,
    props.className,
  );

  return (
    <Link className={linkClassName} href={href}>
      {children}
    </Link>
  );
}

export interface NavItemProps {
  children: string;
  className?: string;
  href: string;
}
