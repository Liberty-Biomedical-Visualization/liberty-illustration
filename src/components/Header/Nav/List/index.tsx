import resolveClassNames from "@/lib/resolveClassNames";

import ListItem, { NavListItemProps as ItemProps } from "./Item";

export default function NavList(props: Readonly<NavListProps>) {
  const { onItemClick, pages } = props;

  const className = resolveClassNames(baseClassNames, props.className);

  const items = pages.map((page, index) => {
    const isLastPage = index === pages.length - 1;
    return transformToItem(page, isLastPage, onItemClick);
  });

  return <ul className={className}>{items}</ul>;
}

export interface NavListProps {
  className?: string;
  onItemClick: ItemProps["onClick"];
  pages: readonly Readonly<PageNavData>[];
}

const baseClassNames = "md:flex";

function transformToItem(
  page: Readonly<PageNavData>,
  isLastPage: boolean,
  onClick: NavListProps["onItemClick"],
) {
  const { href, name } = page;

  const className = isLastPage
    ? ""
    : "md:after:content-pipe md:after:mx-2 md:after:text-foreground";

  return (
    <ListItem className={className} href={href} key={href} onClick={onClick}>
      {name}
    </ListItem>
  );
}

interface PageNavData {
  href: ItemProps["href"];
  name: ItemProps["children"];
}
