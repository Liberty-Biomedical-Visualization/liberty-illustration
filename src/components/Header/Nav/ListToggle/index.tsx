import { faBars, faTimes } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";

import AccessiblyHiddenText from "@/components/AccessiblyHiddenText";
import resolveClassNames from "@/lib/resolveClassNames";

export default function NavListToggle(props: Readonly<NavListToggleProps>) {
  const { id, listIsShown, navListId, onClick } = props;
  const className = resolveClassNames(baseClassName, props.className);

  return (
    <button
      aria-controls={navListId}
      aria-expanded={listIsShown}
      aria-haspopup
      className={className}
      id={id}
      onClick={onClick}
    >
      <FontAwesomeIcon
        fixedWidth
        icon={listIsShown ? faTimes : faBars}
        size="2x"
      />
      <AccessiblyHiddenText>Navigation</AccessiblyHiddenText>
    </button>
  );
}

export interface NavListToggleProps {
  className?: string;
  id: string;
  listIsShown: boolean;
  navListId: string;
  onClick: () => void;
}

const baseClassName =
  "cursor-pointer fixed right-4 md:hidden py-0.5 top-4 z-20";
