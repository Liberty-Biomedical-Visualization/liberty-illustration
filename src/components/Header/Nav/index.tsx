import resolveClassNames from "@/lib/resolveClassNames";

import NavItem, { type NavItemProps } from "./Item";

export default function Nav(props: Readonly<NavProps>) {
  const { pages } = props;

  const items = pages.map((page, index) =>
    transformToNavItem(page, index === pages.length - 1),
  );

  const className = resolveClassNames(
    "flex flex-nowrap flex-row",
    props.className,
  );

  return (
    <nav className={className}>
      <ul className="flex flex-nowrap flex-row">{items}</ul>
    </nav>
  );
}

export interface NavProps {
  className?: string;
  pages: readonly Readonly<PageNavData>[];
}

function transformToNavItem(page: Readonly<PageNavData>, isLastPage: boolean) {
  const { href, name } = page;

  const className = isLastPage
    ? ""
    : // \x22 is the unicode escape character for the double quotation mark. JSDOM
      // is not able to parse raw quotation marks in CSS rules, which causes tests
      // that depend on styling to fail.
      "after:content-[\x22|\x22] after:mx-2 after:text-foreground";

  return (
    <li className={className} key={href}>
      <NavItem href={href}>{name}</NavItem>
    </li>
  );
}

interface PageNavData {
  href: NavItemProps["href"];
  name: NavItemProps["children"];
}
