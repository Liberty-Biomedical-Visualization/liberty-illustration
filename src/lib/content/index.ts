import * as contentful from "contentful";

import packageData from "../../../package.json";
import config from "../config";
import isDefined from "../isDefined";
import { Json, JsonTitle } from "./jsonConfiguration";
import { validateJson } from "./validateJson";

/**
 * Gets `AssetData` by ID.
 *
 * Throws an `Error` if the asset is not found.
 */
export async function getAssetData(id: string) {
  const asset = await client.getAsset(id);
  return mapContentfulAssetToAssetData(asset);
}

export interface AssetData {
  description?: string;
  title?: string;
  url: string;
}

/**
 * Gets `ImageData` by ID.
 *
 * Throws an `Error` if the image is not found, or the asset is not an image.
 */
export async function getImageData(id: string) {
  const asset = await client.getAsset(id);
  return mapContentfulAssetToImageData(asset);
}

export interface ImageData {
  description?: string;
  height: number;
  src: string;
  title?: string;
  width: number;
}

/**
 * Gets an `ImageGallery` by ID.
 *
 * Throws an `Error` if the gallery is not found.
 */
export async function getImageGallery(id: string) {
  const entry = await client.getEntry<ImageGallerySkeleton>(id);
  return mapEntryToImageGallery(entry);
}

export interface ImageGallery {
  images: ImageData[];
  title: string;
}

/**
 * Gets a `JsonReference` by title.
 *
 * Throws an error if the JSON is not found or not properly structured.
 */
export async function getJsonByTitle<K extends JsonTitle>(title: K) {
  const jsons = await client.getEntries<JsonSkeleton<Json<K>>>({
    content_type: "json",
    "fields.title": title,
  });
  const json = jsons.items[0];

  if (json === undefined) {
    throw new Error(`JSON with title "${title}" not found`);
  }

  validateJson(json.fields.value, title);
  return json.fields.value;
}

/**
 * Gets a `VideoGallery` by ID.
 *
 * Throws an `Error` if the gallery is not found.
 */
export async function getVideoGallery(id: string) {
  const entry = await client.getEntry<VideoGallerySkeleton>(id);
  return mapEntryToVideoGallery(entry);
}

export interface VideoGallery {
  title: string;
  videos: VideoData[];
}

export interface VideoData {
  title: string;
  url: string;
}

function mapContentfulAssetToAssetData(asset: ContentfulAsset) {
  const { description, file, title } = asset.fields;

  if (file === undefined) {
    throw new Error("file is undefined");
  }

  const assetData: AssetData = { url: "https:" + file.url };

  if (description) {
    assetData.description = description;
  }

  if (title) {
    assetData.title = title;
  }

  return assetData;
}

function mapContentfulAssetToImageData(asset: ContentfulAsset) {
  const { file } = asset.fields;

  if (file === undefined) {
    throw new Error("file is undefined");
  }

  const { image } = file.details;

  if (image === undefined) {
    throw new Error("image is undefined");
  }

  const { height, width } = image;
  const { description, title, url } = mapContentfulAssetToAssetData(asset);
  const imageData: ImageData = { height, src: url, width };

  if (description) {
    imageData.description = description;
  }

  if (title) {
    imageData.title = title;
  }

  return imageData;
}

type ContentfulAsset = contentful.Asset<"WITHOUT_UNRESOLVABLE_LINKS", string>;

function mapEntryToImageGallery(entry: ImageGalleryEntry) {
  const { images, title } = entry.fields;
  const mappedImages = images
    .filter(isDefined)
    .map(mapContentfulAssetToImageData);
  const gallery: ImageGallery = { images: mappedImages, title };
  return gallery;
}

type ImageGalleryEntry = contentful.Entry<
  ImageGallerySkeleton,
  "WITHOUT_UNRESOLVABLE_LINKS",
  string
>;

interface ImageGallerySkeleton {
  contentTypeId: "imageGallery";
  fields: {
    images: contentful.EntryFieldTypes.Array<contentful.EntryFieldTypes.AssetLink>;
    title: contentful.EntryFieldTypes.Text;
  };
}

function mapEntryToVideoGallery(entry: VideoGalleryEntry) {
  const { title, videos } = entry.fields;
  const mappedVideos = videos.filter(isDefined).map((video) => video.fields);
  const gallery: VideoGallery = { title, videos: mappedVideos };
  return gallery;
}

type VideoGalleryEntry = contentful.Entry<
  VideoGallerySkeleton,
  "WITHOUT_UNRESOLVABLE_LINKS",
  string
>;

interface VideoGallerySkeleton {
  contentTypeId: "imageGallery";
  fields: {
    title: contentful.EntryFieldTypes.Text;
    videos: contentful.EntryFieldTypes.Array<
      contentful.EntryFieldTypes.EntryLink<VideoLinkSkeleton>
    >;
  };
}

interface VideoLinkSkeleton {
  contentTypeId: "videoLink";
  fields: {
    title: contentful.EntryFieldTypes.Text;
    url: contentful.EntryFieldTypes.Text;
  };
}

interface JsonSkeleton<T extends JsonReference> {
  contentTypeId: "json";
  fields: {
    value: contentful.EntryFieldTypes.Object<T>;
    title: contentful.EntryFieldTypes.Text;
  };
}

type JsonReference = JsonArray | JsonObject;
type JsonArray = JsonValue[];
type JsonObject = { [key: string]: JsonValue };
type JsonValue = JsonArray | JsonObject | boolean | null | number | string;

const { dependencies, name, version } = packageData;

const client = contentful.createClient({
  accessToken: config.CONTENTFUL_ACCESS_TOKEN,
  application: `${name}/${version}`,
  environment: config.CONTENTFUL_ENVIRONMENT,
  integration: `next/${dependencies.next}`,
  space: config.CONTENTFUL_SPACE_ID,
}).withoutUnresolvableLinks;
