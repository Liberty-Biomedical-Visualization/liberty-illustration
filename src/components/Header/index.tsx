import resolveClassNames from "@/lib/resolveClassNames";

import Brand, { type BrandProps } from "./Brand";

export default function Header(props: HeaderProps) {
  const { logoData } = props;
  const className = resolveClassNames(
    "flex flex-col items-center",
    props.className,
  );

  return (
    <header className={className}>
      <Brand logoData={logoData} />
    </header>
  );
}

export interface HeaderProps {
  className?: string;
  logoData: BrandProps["logoData"];
}
