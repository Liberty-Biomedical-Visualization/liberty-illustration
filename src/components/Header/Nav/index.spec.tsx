import { describe, expect, it } from "@jest/globals";
import { screen } from "@testing-library/dom";

import createTestRender from "@/lib/createTestRender";

import Nav from ".";

describe("Nav", () => {
  it("should display navigation", () => {
    renderNav();
    const result = screen.getByRole("navigation");
    expect(result).toBeVisible();
  });

  it("should display a list of links from passed pages", () => {
    const pages = [
      { href: "/", name: "Home" },
      { href: "/portfolio", name: "Portfolio" },
    ];
    renderNav({ pages });

    const listItems = screen.getAllByRole("listitem");

    const homeLink = screen.getByRole("link", { name: "Home" });
    const containsHomeLink = listItems.some((listItem) =>
      listItem.contains(homeLink),
    );
    expect(containsHomeLink).toBe(true);

    const portfolioLink = screen.getByRole("link", { name: "Portfolio" });
    const containsPortfolioLink = listItems.some((listItem) =>
      listItem.contains(portfolioLink),
    );
    expect(containsPortfolioLink).toBe(true);
  });

  it("should pass className to the outermost element", () => {
    const className = "foo";
    const result = renderNav({ className });
    expect(result.container.firstChild).toHaveClass(className);
  });

  it("should match the snapshot", () => {
    const result = renderNav();
    expect(result.container).toMatchSnapshot();
  });
});

const renderNav = createTestRender(Nav, {
  pages: [],
});
