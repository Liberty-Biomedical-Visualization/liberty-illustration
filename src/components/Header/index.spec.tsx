import { describe, expect, it } from "@jest/globals";
import { screen } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";

import { describedImage, minimalImage } from "@/lib/content/test-image-data";
import createTestRender from "@/lib/createTestRender";

import Header from ".";

describe("Header", () => {
  it("should display the site banner", () => {
    renderHeader();
    const result = screen.getByRole("banner");
    expect(result).toBeVisible();
  });

  it("should display an accessible logo when passed image data with a description", () => {
    renderHeader({ logoData: describedImage });
    const result = screen.getByRole("img", {
      name: describedImage.description,
    });
    expect(result).toBeVisible();
  });

  it("should display a list of links from passed pages when nav list toggle is clicked", async () => {
    const pages = [
      { href: "/", name: "Home" },
      { href: "/portfolio", name: "Portfolio" },
    ];
    renderHeader({ pages });

    const user = userEvent.setup();
    const listToggle = screen.getByRole("button", { name: "Navigation" });
    await user.click(listToggle);

    const list = screen.getByRole("list");
    expect(list).toBeVisible();

    const listItems = screen.getAllByRole("listitem");

    const homeLink = screen.getByRole("link", { name: "Home" });
    const containsHomeLink = listItems.some((listItem) =>
      listItem.contains(homeLink),
    );
    expect(containsHomeLink).toBe(true);

    const portfolioLink = screen.getByRole("link", { name: "Portfolio" });
    const containsPortfolioLink = listItems.some((listItem) =>
      listItem.contains(portfolioLink),
    );
    expect(containsPortfolioLink).toBe(true);
  });

  it("should pass className to the outermost element", () => {
    const className = "foo";
    const result = renderHeader({ className });
    expect(result.container.firstChild).toHaveClass(className);
  });

  it("should match the snapshot", () => {
    const result = renderHeader();
    expect(result.container).toMatchSnapshot();
  });
});

const renderHeader = createTestRender(Header, {
  disableTabbableDisplayCheck: true,
  logoData: minimalImage,
  pages: [],
});
