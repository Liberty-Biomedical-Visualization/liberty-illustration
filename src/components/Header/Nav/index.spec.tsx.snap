// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav should match the snapshot 1`] = `
<div>
  <nav
    class="flex flex-nowrap flex-row"
  >
    <ul
      class="flex flex-nowrap flex-row"
    />
  </nav>
</div>
`;
